src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -976,4 +977,18 @@ public ProjectState UpdateDocumentsOrder(ImmutableList<DocumentId> documentIds)
                 ? CreateLazyLatestDocumentTopLevelChangeVersion(newDocument, newDocumentStates, newAdditionalDocumentStates)
                 : _lazyLatestDocumentTopLevelChangeVersion;
     }
+
+    public void AddDocumentIdsWithFilePath(ref TemporaryArray<DocumentId> temporaryArray, string filePath)
+    {
+        this.DocumentStates.AddDocumentIdsWithFilePath(ref temporaryArray, filePath);
+        this.AdditionalDocumentStates.AddDocumentIdsWithFilePath(ref temporaryArray, filePath);
+        this.AnalyzerConfigDocumentStates.AddDocumentIdsWithFilePath(ref temporaryArray, filePath);
+    }
+
+    public DocumentId? GetFirstDocumentIdWithFilePath(string filePath)
```

Comment:
A new helper for the common case of adding a file.  We only need to find the first doc with that filepath to sync our contents to.  this has a few major benefits:

1. it avoids unnecessary array allocs
2. it avoids returning the same doc that we're looking for related docs for.
3. it means we don't create the filepathmap for a project when adding a doc to it, only to throw it away because we're adding a doc to it :)